### DISCUSSION TITLE
How to make the table of contents show headers added with React in an MDX file?

### DISCUSSION THREAD

**Discussion by albertothedev** [facebook/docusaurus#8974] [Q&A] [answered]
This code will render two `h2` headers, yet only the one directly specified is shown in the table of contents.

`docs/example.mdx` 
![Screenshot_20230513_163919](https://github.com/facebook/docusaurus/assets/46131317/d6099d8a-fef1-47f1-be66-c3a37903b3d2)

Website
![Screenshot_20230513_163849](https://github.com/facebook/docusaurus/assets/46131317/74095169-731c-4ee6-8d38-47558aaf2345)

**Comment by Josh-Cena** [comment, marked as answer]
Please see https://github.com/facebook/docusaurus/issues/3915

**Reply by PaulRBerg** [reply]
I have the same problem as @albertothedev.

I've read through the issue you linked, but I am still unsure whether this feature (ToC headers when generating MDX content dynamically with React) is supported or not.

Could you please provide a summary?

cc @slorber

**Reply by slorber** [reply]
We don't support that because our remark plugin that extracts headings works at compile time.

The JSX `<h2>` tag might be produced by a very deeply nested component. The React code you provide isn't evaluated, we don't run it to see what gets rendered, we extract it at compile time and need an easy way to do so.

If you want full control over the TOC, you have `export const toc = []` and create your own TOC items manually.


---

In the future, the best we could do is to eventually support this use-case better by allowing your custom React components to expose a "toc slice".

```tsx
export default function MyComp() {
  return <h2>Hey!</h2>
}

MyComp.toc = [{level: 2, value: "Hey!", id: "hey"}]
```
---

Various related discussions are in https://github.com/facebook/docusaurus/issues/3915
